daemons/lib-copy-offload/copy-offload.c

```diff
@@ -0,0 +1,219 @@
+    if(res != CURLE_OK) {
+        // Things that are not HTTP response status codes.
+        // See libcurl-errors(3) and the example in CURLOPT_ERRORBUFFER(3).
+        size_t len = strlen(errbuf);
+        if (len)
+            snprintf(offload->err_message, COPY_OFFLOAD_MSG_SIZE, "libcurl: (%d) %s%s", res, errbuf, ((errbuf[len - 1] != '\n') ? "\n" : ""));
+        else
+            snprintf(offload->err_message, COPY_OFFLOAD_MSG_SIZE, "libcurl: (%d) %s\n", res, curl_easy_strerror(res));
+    } else {
+        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
+        offload->http_code = http_code;
+    }
+
+    return http_code;
+}
+
+/* List the active copy-offload requests.
+ * The caller is responsible for calling free() on @output if *output is non-NULL.
+ */
+int copy_offload_list(COPY_OFFLOAD *offload, char **output) {
+    long http_code;
+    struct memory chunk = {NULL, 0};
+    int ret = 1;
+    char urlbuf[COPY_OFFLOAD_URL_SIZE];
+
+    snprintf(urlbuf, COPY_OFFLOAD_URL_SIZE, "%s/list", *offload->host_and_port);
+    curl_easy_setopt(offload->curl, CURLOPT_URL, urlbuf);
+
+    http_code = copy_offload_perform(offload, &chunk);
+    if (http_code == 200) {
+        if (chunk.response != NULL) {
+            *output = strdup(chunk.response);
+            // chop
+            char *newline = strchr(*output, '\n');
+            *newline = '\0';
+            free(chunk.response);
+        }
+        ret = 0;
+    }
+    return ret;
+}
+
+/* Cancel a specific copy-offload request. */
+int copy_offload_cancel(COPY_OFFLOAD *offload, char *job_name) {
+    long http_code;
+    struct memory chunk = {NULL, 0};
+    int ret = 1;
+    char urlbuf[COPY_OFFLOAD_URL_SIZE];
+
+    snprintf(urlbuf, COPY_OFFLOAD_URL_SIZE, "%s/cancel/%s", *offload->host_and_port, job_name);
+    curl_easy_setopt(offload->curl, CURLOPT_URL, urlbuf);
+    curl_easy_setopt(offload->curl, CURLOPT_CUSTOMREQUEST, "DELETE");
+
+    http_code = copy_offload_perform(offload, &chunk);
+    if (http_code == 204) {
+        ret = 0;
+    } else if (chunk.response != NULL) {
+        free(chunk.response);
+    }
+    return ret;
+}
+
+/* Submit a new copy-offload request.
+ * The caller is responsible for calling free() on @job_name if *job_name is non-NULL.
+ */
+int copy_offload_copy(COPY_OFFLOAD *offload, char *compute_name, char *workflow_name, char *source_path, char *dest_path, char **job_name) {
+    long http_code;
+    struct memory chunk = {NULL, 0};
+    int ret = 1;
+    char urlbuf[COPY_OFFLOAD_URL_SIZE];
+    char postbuf[COPY_OFFLOAD_POST_SIZE];
+
+    snprintf(urlbuf, COPY_OFFLOAD_URL_SIZE, "%s/trial", *offload->host_and_port);
+    curl_easy_setopt(offload->curl, CURLOPT_URL, urlbuf);
+
+    char *offload_req = "{\"computeName\": \"%s\", \"workflowName\": \"%s\", \"sourcePath\": \"%s\", \"destinationPath\": \"%s\", \"dmProfile\": \"copy-offload-nonmpi\", \"dryrun\": true, \"dcpOptions\": \"\", \"logStdout\": true, \"storeStdout\": false}";
```

Comment:
For the profile, are we going to require users to have separate profiles that have the correct mkdir/stat command lines?

Reply:
They will have to refer to the correct profile, like I did in the stuff I showed earlier today.
This is how the current implementation, using the nnf-dm daemon, is doing things as well.